google-cloud-spanner/src/main/java/com/google/cloud/spanner/Interval.java

```diff
@@ -0,0 +1,316 @@
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT, "Invalid Interval String: " + interval);
+    }
+
+    long years = Long.parseLong(getNullOrDefault(matcher, 1).replace("Y", ""));
+    long months = Long.parseLong(getNullOrDefault(matcher, 2).replace("M", ""));
+    long days = Long.parseLong(getNullOrDefault(matcher, 3).replace("D", ""));
+    long hours = Long.parseLong(getNullOrDefault(matcher, 5).replace("H", ""));
+    long minutes = Long.parseLong(getNullOrDefault(matcher, 6).replace("M", ""));
+    BigDecimal seconds = new BigDecimal(getNullOrDefault(matcher, 7).replace("S", ""));
+
+    long totalMonths = years * MONTHS_PER_YEAR + months;
+    BigInteger totalNanos = seconds.movePointRight(9).toBigInteger();
+    totalNanos =
+        totalNanos.add(BigInteger.valueOf(minutes * SECONDS_PER_MINUTE).multiply(NANOS_PER_SECOND));
+    totalNanos =
+        totalNanos.add(BigInteger.valueOf(hours * SECONDS_PER_HOUR).multiply(NANOS_PER_SECOND));
+
+    BigInteger totalMicros = totalNanos.divide(BigInteger.valueOf(NANOS_PER_MICRO));
+    BigInteger nanoFractions =
+        totalNanos.subtract(totalMicros.multiply(BigInteger.valueOf(NANOS_PER_MICRO)));
+
+    return Interval.builder()
+        .setMonths(Math.toIntExact(totalMonths))
+        .setDays(Math.toIntExact(days))
+        .setMicros(totalMicros.longValueExact())
+        .setNanoFractions(nanoFractions.shortValueExact())
+        .build();
+  }
+
+  /** Converts Interval to ISO8601 Duration Formatted String. */
+  public String ToISO8601() {
+    StringBuilder result = new StringBuilder();
+    result.append("P");
+
+    long months = this.months();
+    long years = months / MONTHS_PER_YEAR;
+    months = months - years * MONTHS_PER_YEAR;
+
+    if (years != 0) {
+      result.append(String.format("%dY", years));
+    }
+
+    if (months != 0) {
+      result.append(String.format("%dM", months));
+    }
+
+    if (this.days() != 0) {
+      result.append(String.format("%dD", this.days()));
+    }
+
+    BigInteger nanos = this.nanos();
+    BigInteger zero = BigInteger.valueOf(0);
+    if (nanos.compareTo(zero) != 0) {
+      result.append("T");
+      BigInteger hours = nanos.divide(NANOS_PER_HOUR);
+
+      if (hours.compareTo(zero) != 0) {
+        result.append(String.format("%sH", hours));
+      }
+
+      nanos = nanos.subtract(hours.multiply(NANOS_PER_HOUR));
+      BigInteger minutes = nanos.divide(NANOS_PER_MINUTE);
+      if (minutes.compareTo(zero) != 0) {
+        result.append(String.format("%sM", minutes));
+      }
+
+      nanos = nanos.subtract(minutes.multiply(NANOS_PER_MINUTE));
+      BigDecimal seconds = new BigDecimal(nanos).movePointLeft(9);
+
+      if (seconds.compareTo(new BigDecimal(zero)) != 0) {
+        result.append(String.format("%sS", seconds.stripTrailingZeros()));
+      }
+    }
+
+    if (result.length() == 1) {
+      result.append("0Y");
+    }
+
+    return result.toString();
+  }
+
```

Comment:
nit: also add a `toString()` method:

```suggestion
@Override
public String toString() {
  return toISO8601();
}
```

Reply:
added.